src/gsi-ncdiag/radiances.py

```diff
@@ -1,222 +1,245 @@
+                var_out[...] = vdata
+            else:
+                pass
+        ncout.close()
+
+    def toIODAobs(self, OutDir, clobber=True):
+        """ toIODAobs(OutDir,clobber=True)
+     output observations from the specified GSI diag file
+     to the JEDI/IODA observation format
+        """
+        import ioda_conv_ncio as iconv
+        import os
+        from collections import defaultdict
+        import rad_dicts as rd
+        import numpy as np
+        import datetime as dt
+        # set up a NcWriter class
+        outname = OutDir + '/' + self.sensor + '_' + self.satellite + \
+            '_obs_' + self.validtime.strftime("%Y%m%d%H") + '.nc4'
+        if not clobber:
+            if (os.path.exists(outname)):
+                print("File exists. Skipping and not overwriting:")
+                print(outname)
+                return
+        RecKeyList = []
+        LocKeyList = []
+        LocVars = []
+        AttrData = {}
+        varDict = defaultdict(lambda: defaultdict(dict))
+        outdata = defaultdict(lambda: defaultdict(dict))
+        rec_mdata = defaultdict(lambda: defaultdict(dict))
+        loc_mdata = defaultdict(lambda: defaultdict(dict))
+        var_mdata = defaultdict(lambda: defaultdict(dict))
+        # get list of location variable for this var/platform
+        for ncv in self.df.variables:
+            if ncv in rd.LocKeyList:
+                LocKeyList.append(rd.LocKeyList[ncv])
+                LocVars.append(ncv)
+        # LocKeyList.append(('ObsIndex','integer')) # to ensure unique obs
+        # for now, record len is 1 and the list is empty?
+        recKey = 0
+        writer = iconv.NcWriter(outname, RecKeyList, LocKeyList)
+
+        chan_number = self.df['sensor_chan'][:]
+        chan_number = chan_number[chan_number >= 0]
+        chan_indx = self.df['Channel_Index'][:]
+        nchans = len(chan_number)
+        nlocs = self.nobs / nchans
+        chanlist = chan_indx[:nchans]
+        for a in chanlist:
+            value = "brightness_temperature_{:d}".format(a)
+            varDict[value]['valKey'] = value, writer.OvalName()
+            varDict[value]['errKey'] = value, writer.OerrName()
+            varDict[value]['qcKey'] = value, writer.OqcName()
+
+        obsdata = self.df['Observation'][:]
+        obserr = self.df['error_variance'][:]
+        obsqc = self.df['QC_Flag'][:]
+        gsivars = rd.gsi_add_vars
+
+        # loop through channels for subset
+        var_names = []
+        for c in range(len(chanlist)):
+            value = "brightness_temperature_{:d}".format(chanlist[c])
+            var_names.append(value)
+            print(self.obstype, value)
+            idx = chan_indx == chanlist[c]
+            if (np.sum(idx) == 0):
+                print("No matching observations for:")
+                print(value)
+                continue
+            obsdatasub = obsdata[idx]
+            obsdatasub[obsdatasub>1e6] = np.nan
```

Comment:
Presently the large IODA NC obs files for amsua have BT values of exactly 1e6.  Will this catch those?

Reply:
No, I've changed it to be 9e5 as that should still be sufficiently large.

Reply:
Thank you Cory.

Reply:
No problem, thanks for double checking this.

Reply:
@jjguerrette using np.nan does not work with the ctest, as it can't properly handle NaN or Inf. I'm using the default netCDF fill value now of 9.96921e+36, hopefully this is fine.

Reply:
That should work since @svahl991 told me that any ObsValue larger than some predefined value (1e+10 maybe, please give the correct value) is treated as missing in IODA.  I assume that is why we already have values of 1e+11 in the radiance files.

Reply:
I just did some grepping and confirmed that the threshold value is 1e+8, which is what I thought. So any value larger than that is treated as a missing value. One thing to note, though, is that this threshold value is defined in the NetcdfIO source code. What this means is that it's defined in the code that reads the IODA data from IODA netCDF files. Right now we only use netCDF files to store IODA data so the distinction isn't important, but if we ever used a different storage medium (like ODC), we would use a different I/O class, and there's no mechanism to enforce that the same value is used by other I/O classes, or even that any threshold is set at all. We should probably set this threshold at a higher, public, level, so that it's not quite so hidden, and enforce somehow that all IODA I/O classes use the same value. I think I'll enter a Zenhub issue regarding this.